zsl/interface/celery/worker.py

```diff
@@ -0,0 +1,108 @@
+"""
+from __future__ import unicode_literals
+
+import sys
+
+from celery import Celery, shared_task
+from injector import Module, singleton, provides
+
+from zsl import Config
+from zsl.utils.injection_helper import inject
+from zsl.interface.task_queue import TaskQueueWorker
+from zsl.task.job_context import Job
+
+
+class CeleryTaskQueueWorkerBase(TaskQueueWorker):
+    """Base class for celery task queue worker.
+
+    It initialize the celery app in constructor.
+    """
+
+    @inject(config=Config)
+    def __init__(self, config):
+        # type: (Config) -> None
+        super(CeleryTaskQueueWorkerBase, self).__init__()
+
+        self.celery_app = Celery()
+        self.celery_app.config_from_object(config['CELERY'])
+
+    def execute_celery_task(self, job_data):
+        # type: (dict) -> dict
+        """Creates job from task and executes the job.
+
+        :param job_data: job's data
+        :return: job results
+        :rtype: dict
+        """
+        job = Job(job_data)
+
+        return self.execute_job(job)
+
+
+class CeleryTaskQueueOutsideWorker(CeleryTaskQueueWorkerBase):
+    """Celery worker used only for task execution.
+
+    This can be used when the worker is manage with some other application,
+    like `celery worker` or `celery multi`.
+    """
+
+    def stop_worker(self):
+        self._app.logger.error("Running from celery worker, kill from shell!")
+
+    def run(self):
+        self._app.logger.error("Running from celery worker, start from shell!")
+
+
+class CeleryTaskQueueMainWorker(CeleryTaskQueueWorkerBase):
+    """Worker implementation for Celery task queue.
+    """
+    def __init__(self):
+        super(CeleryTaskQueueMainWorker, self).__init__()
+
+        self.celery_worker = None
+
+    def stop_worker(self):
+        self._app.logger.info("Stopping Celery worker on demand - quitting.")
+        self.celery_worker.stop()
+
+    def run(self, argv, *args, **kwargs):
+        self._app.logger.info("Running the worker.")
+        self.celery_worker = self.celery_app.worker_main((sys.argv[0],) + argv)
+
+
+class CeleryTaskQueueMainWorkerModule(Module):
+    """Injector module for celery main worker."""
+    @singleton
+    @provides(CeleryTaskQueueWorkerBase)
+    def provide_worker(self):
+        return CeleryTaskQueueMainWorker()
+
+
+class CeleryTaskQueueOutsideWorkerModule(Module):
+    """Injector module for celery outside worker."""
+    @singleton
+    @provides(CeleryTaskQueueWorkerBase)
+    def provide_worker(self):
+        return CeleryTaskQueueMainWorker()
+
+
+@shared_task
+@inject(worker=CeleryTaskQueueWorkerBase)
+def zsl_task(job_data, worker):
+    # type: (dict, CeleryTaskQueueWorkerBase) -> dict
+    """This function will be registered as a celery task.
+
+    :param job_data: task data
+    :param worker: current celery worker
+    :type worker: CeleryTaskQueueWorkerBase
+    :return: task results
+    """
+    return worker.execute_celery_task(job_data)
```

Comment:
For now the task execution is tied with the worker instance itself, I used injection to get the current, single instance. The problem with this is, that the worker creates its own celery app instance. I can imagine a use-case, where the celery app creation is handled in the client and `zsl_task` is imported and registred properly.

```python
from celery import Celery
from zsl.interface.gearman import zsl_task
from my_old_app import all_the_tasks

# .. code code code ..

app = Celery()
```
This would integrate nice into other projects, but it is only possible if the celery app creation is left to the zsl worker.